homeassistant/components/hue/config_flow.py

```diff
@@ -194,26 +194,28 @@ async def async_step_ssdp(self, discovery_info: ssdp.SsdpServiceInfo) -> FlowRes
         """
         # Filter out non-Hue bridges #1
         if (
-            discovery_info.get(ssdp.ATTR_UPNP_MANUFACTURER_URL)
+            discovery_info.upnp.get(ssdp.ATTR_UPNP_MANUFACTURER_URL)
             not in HUE_MANUFACTURERURL
         ):
             return self.async_abort(reason="not_hue_bridge")
 
         # Filter out non-Hue bridges #2
         if any(
-            name in discovery_info.get(ssdp.ATTR_UPNP_FRIENDLY_NAME, "")
+            name in discovery_info.upnp.get(ssdp.ATTR_UPNP_FRIENDLY_NAME, "")
             for name in HUE_IGNORED_BRIDGE_NAMES
         ):
             return self.async_abort(reason="not_hue_bridge")
 
-        if (
-            not discovery_info.get(ssdp.ATTR_SSDP_LOCATION)
-            or ssdp.ATTR_UPNP_SERIAL not in discovery_info
-        ):
+        if ssdp.ATTR_UPNP_SERIAL not in discovery_info.upnp:
+            return self.async_abort(reason="not_hue_bridge")
+
+        host = urlparse(discovery_info.ssdp_location).hostname
+        if not host:
             return self.async_abort(reason="not_hue_bridge")
 
-        host = urlparse(discovery_info[ssdp.ATTR_SSDP_LOCATION]).hostname
-        bridge = await self._get_bridge(host, discovery_info[ssdp.ATTR_UPNP_SERIAL])
+        bridge = await self._get_bridge(
+            str(host), discovery_info.upnp[ssdp.ATTR_UPNP_SERIAL]
```

Comment:
```suggestion
            host, discovery_info.upnp[ssdp.ATTR_UPNP_SERIAL]
```

Reply:
That was my initial thought also, but `hostname` return `str | bytes | None`

Reply:
huh ? why is that ? Is that some old python2 compatability stuff ?
On my setup it returns either str or None

![Schermafbeelding 2021-11-30 om 11 08 47](https://user-images.githubusercontent.com/6389780/144027501-03943edc-443f-42cd-9c5f-76379b2fb52f.png)

Reply:
hmm, my dev machine is on python 3.10... might be that urlparse changed between py versions...
I think this is why I added the type: i_gnore[arg-type]_ here...

Reply:
Looking at the docs the namedtuple Paresult either have a str or None value so not bytes.
https://docs.python.org/3/library/urllib.parse.html

Reply:
If you prefer that I use the type ignore, then I can remove the cast and add `# type: ignore[arg-type]`
But I thought it was better to not use the type ignore.

Reply:
> hostname returns AnyStr which is TypeVar('AnyStr', bytes, str)

Strange, not on my setup. Typehint is Str | None and so is the documentation.
is the hostname is None guard is there, the type: ignore can be dropped. Can you try what mypy thinks of that on CI ? Still interesting to know what python version you're running to explain the difference we're seeing.

If the returntype is going to change to AnyStr in upcoming release we could already anticipate on that and change the argument of the get_bridge to AnyStr too and handle conversion there.

Reply:
I am running Python 3.9.

`if not host:` ensures that `None` is filtered out, so I don't think changing that is useful.

I've just now pushed an update but I had to use `--no-verify`

Reply:
> 

correct but the explicit check for None makes it more clear to readers of your code that you're handling the situation where the urlparse attribute hostname is None.

Reply:
OK. I've updated again.
However the issue with host being `str | bytes` is still there:
https://github.com/home-assistant/core/runs/4366467027?check_suite_focus=true

And if you look at your screenshot above it does show `AnyStr`.
This is what I have on my machine.
![image](https://user-images.githubusercontent.com/6771947/144038777-2112cf4e-cc24-4c72-8d40-0e186455509f.png)

Reply:
@marcelveldt did you try to remove the `# type: ignore[arg-type]` that you appear to have in your code?
It's not in the current code and I think it is conflicting with this discussion.

Reply:
Aha! I think I found the root cause... I have a similar issue in `samsungtv` PR #60595
- previously `discovery_info[ATTR_SSDP_LOCATION]` return type was `Any`
- now `discovery_info.ssdp_location` is defined as `str | None`

If the first argument of urlparse is `Any`, then `urlparse.hostname` returns `str | None`
If the first argument of urlparse is `str | None`, then `urlparse.hostname` returns `str | bytes | None`

I'll see if I can clean this up better.

Reply:
You are 100% right. The documentation doesn't match wil reality and the type is indeed AnyStr, yikes.
Maybe add a very minor extra line that converts to string if the output is bytes ?
Or indeed leave your forced to string conversion in place.

Reply:
I have reworked this, and added an extra test for an invalid hostname.

Reply:
Looks good. Mypy happy now ?

Reply:
Yes the tests pass. I think this is good to merge.